cranelift/isle/isle/src/serialize.rs

```diff
@@ -0,0 +1,823 @@
+                debug_assert_eq!(base_state, BindingState::Available);
+                self.use_expr(source);
+                self.scope.ready[source.index()] = base_state;
+                self.add_bindings();
+
+                let mut child = self.new_block();
+                child.set_ready(source, BindingState::Matched);
+                child.set_ready(result, BindingState::Emitted);
+                child.add_bindings();
+                let body = child.sort(order);
+                ControlFlow::Loop { result, body }
+            }
+        }
+    }
+
+    /// Advance the given binding to a new state. The new state usually should
+    /// be greater than the existing state; but at the least it must never
+    /// go backward.
+    fn set_ready(&mut self, source: BindingId, state: BindingState) {
+        let old = &mut self.scope.ready[source.index()];
+        debug_assert!(*old <= state);
+
+        // Add candidates for this binding, but only when it first becomes
+        // available.
+        if let BindingState::Unavailable = old {
+            // A binding site can't have all of these kinds of constraint,
+            // and many have none. But `best_control_flow` has to check all
+            // candidates anyway, so let it figure out which (if any) of these
+            // are applicable. It will only check false candidates once on any
+            // partition, removing them from this list immediately.
+            self.scope.candidates.extend([
+                Candidate::new(HasControlFlow::Match(source)),
+                Candidate::new(HasControlFlow::Loop(source)),
+            ]);
+            self.scope
+                .equal_candidates
+                .push(EqualCandidate::new(source));
+        }
+
+        *old = state;
+    }
+
+    /// For the specified set of rules, heuristically choose which control-flow
+    /// will minimize redundant work when the generated code is running.
+    fn best_control_flow(&mut self, order: &mut [usize]) -> Option<HasControlFlow> {
+        // If there are no rules left, none of the candidates will match
+        // anything in the `retain_mut` call below, so short-circuit it.
+        if order.is_empty() {
+            // This is only read in a debug-assert but it's fast so just do it
+            self.scope.candidates.clear();
+            return None;
+        }
+
+        let mut set_score = |score: &mut Score, kind| {
+            // This binding's state may have changed since we last looked at it.
+            let source = match kind {
+                HasControlFlow::Match(source) => source,
+                HasControlFlow::Equal(source, _) => source,
+                HasControlFlow::Loop(source) => source,
+            };
+            score.state = self.scope.ready[source.index()];
+
+            // Candidates which have already been matched in this partition
+            // must not be matched again. (Note that equality constraints are
```

Comment:
Is this because that would generate dead code?

Reply:
Good question. No, I was concerned about the codegen going into an infinite loop of trying to match the same binding site repeatedly. The first time, it would reduce the partition to only those rules that all have the same pattern for that binding site. Matching the same binding site again within that partition won't move any rules out of the partition and it won't make any more binding sites available, so we'd come back to the same evaluation point without making any progress.

The other argument for this is that there's never anything to be gained from matching a binding site when you're in an evaluation path where you already know exactly what pattern that binding site matches. But the full justification for that is a little complicated because you have to reason about both fall-through cases and equality constraints. It's probably worth writing that argument out in some more detail, but the thought of doing so is making me tired :sweat_smile: